src/passes/OptimizeInstructions.cpp

```diff
@@ -205,6 +210,176 @@ struct OptimizeInstructions
+        //     (i32.sub
+        //       (i32.const 0)
+        //       X
+        //     )
+        //   )
+        // =>
+        //   (i32.sub
+        //     Y
+        //     X
+        //   )
+        Expression* y;
+        Binary* sub;
+        if (matches(curr,
+                    binary(AddInt32,
+                           any(&y),
+                           binary(&sub, SubInt32, i32(0), any())))) {
+          return build(binary(&sub, SubInt32, any(&y), any()));
+        }
+      }
+      {
+        // try de-morgan's AND law,
+        //  (eqz X) and (eqz Y) === eqz (X or Y)
+        // Note that the OR and XOR laws do not work here, as these
+        // are not booleans (we could check if they are, but a boolean
+        // would already optimize with the eqz anyhow, unless propagating).
+        // But for AND, the left is true iff X and Y are each all zero bits,
+        // and the right is true if the union of their bits is zero; same.
+        Unary* un;
+        Binary* bin;
+        Expression *x, *y;
+        if (matches(curr,
+                    binary(&bin,
+                           AndInt32,
+                           unary(&un, EqZInt32, any(&x)),
+                           unary(EqZInt32, any(&y))))) {
+          return build(
+            unary(&un, EqZInt32, binary(&bin, OrInt32, any(&x), any(&y))));
+        }
+      }
+      // Selects
+      if (auto* select = curr->dynCast<Select>()) {
+        select->condition = optimizeBoolean(select->condition);
+      }
+      {
+        // Constant condition, we can just pick the right side (barring side
+        // effects)
+        Expression *ifTrue, *ifFalse;
+        if (matches(curr, select(any(&ifTrue), any(&ifFalse), i32(0)))) {
+          if (!effects(ifTrue).hasSideEffects()) {
+            return ifFalse;
+          } else {
+            return builder.makeSequence(builder.makeDrop(ifTrue), ifFalse);
+          }
+        }
+        if (matches(curr, select(any(&ifTrue), any(&ifFalse), constant()))) {
+          // The condition must be non-zero
+          if (!effects(ifFalse).hasSideEffects()) {
+            return ifTrue;
+          } else {
+            // Don't bother - we would need to reverse the order using a temp
+            // local, which would be bad
+          }
+        }
+      }
+      {
+        // Flip select to remove eqz if we can reorder
+        Select* s;
+        Expression *ifTrue, *ifFalse, *c;
+        if (matches(
+              curr,
+              select(
+                &s, any(&ifTrue), any(&ifFalse), unary(EqZInt32, any(&c)))) &&
+            !effects(ifTrue).invalidates(effects(ifFalse))) {
+          curr = build(select(&s, any(&ifFalse), any(&ifTrue), any(&c)));
+        }
+      }
+      {
+        // Simplify selects between 0 and 1
+        Expression* c;
+        bool matchReversed = matches(curr, select(i32(0), i32(1), any(&c))) ||
+                             matches(curr, select(i64(0), i64(1), any(&c)));
```

Comment:
is it possible unify made `isize` type which unify `i32` and `i64`?
```cpp
struct isize {
  int64_t val;
  isize(int64_t val) : val(val) {}
  bool matches(Expression* expr) const {
    auto* c = expr->dynCast<Const>();
    return c && (c->type == Type::i32 || c->type == Type::i64) && 
           c->value.getInteger() == val;
  }
};
```
and later:
```cpp
bool matchReversed = matches(curr, select(isize(0), isize(1), any(&c)));
```

Reply:
Good idea. I added an `ival` matcher for this case, but kept the `i32` and `i64` matchers as well because they're slightly easier to read. In general I'm trying to follow the DAMP principle ("Descriptive and Meaningful Phrases") rather than the DRY principle in the matcher API.

Reply:
Yes, I don't mean remove i32 and i64, just add unified `isize` / `ival` for such cases